### PR TITLE
TCK3 LPS-76715_2

### PR DESCRIPTION
**Pull request by ChiZelin**
Previous CI result: https://github.com/ChiZelin/liferay-portal/pull/89

### CODE REVIEW

File: portal-impl/src/com/liferay/portal/extraconfig/ExtraPortletAppConfig.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+package com.liferay.portal.extraconfig;
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * @author Leon Chi
+ */
+public class ExtraPortletAppConfig {
+	public ExtraPortletAppConfig(Map<String, String> localeEncodingMappingList){
```
Comment (dantewang) [inline portal-impl/src/com/liferay/portal/extraconfig/ExtraPortletAppConfig.java]: SF: Please insert an empty line before line 24.

File: portal-impl/src/com/liferay/portal/service/impl/PortletLocalServiceImpl.java [hunk 2]
```diff
@@ -2716,6 +2722,39 @@ private void _checkValidPortletId(String portletId)
...
 	}
 
+	private void _readWebXML(String xml, String servletContextName)
+		throws Exception {
+
+		Document document = UnsecureSAXReaderUtil.read(xml);
+
+		Element rootElement = document.getRootElement();
+
+		for (Element localeEncodingMappingListElement :
```
Comment (dantewang) [inline portal-impl/src/com/liferay/portal/service/impl/PortletLocalServiceImpl.java]: Do we really need a for-loop here? Please take a look at the DTD of web.xml to see if it's possible for a web.xml to have multiple <local-encoding-mapping-list> elements.
If not, remove this for-loop.
If it is possible, then your logic will only successfully read the last one appearing in the file.

File: portal-impl/src/com/liferay/portlet/ResourceResponseImpl.java [hunk 3]
```diff
@@ -136,7 +143,37 @@ public void setIntHeader(String name, int value) {
 
 	@Override
 	public void setLocale(Locale locale) {
+		if (locale == null) {
+			return;
+		}
+
 		response.setLocale(locale);
+
+		if (_canSetCharacterEncodingCallSetLocale) {
```
Comment (dantewang) [inline portal-impl/src/com/liferay/portlet/ResourceResponseImpl.java]: Is this `_canSetCharacterEncodingCallSetLocale` logic copied from Pluto? This name looks somehow confusing.

**Comment by dantewang** [comment]
@ChiZelin Sick leave today. See my review.

**Comment by ChiZelin** [comment]
@dantewang 
According the Schema of web.xml, the element <locale-encoding-mapping-list> can appear many times. (But in actual use, I think it can only appear once. Maybe it's the inaccuracy of the schema of web.xml definition?)
I will modify the code according to the situation that this element may repeat. OK? 

The name in pluto is ```canSetLocaleEncoding```.  In order to avoid the same as Pluto, I changed the name. Do you have any suggestions?

Thanks!

**Comment by dantewang** [comment]
@ChiZeLin could you send me the link to the web.xml schema?

Leon Chi <notifications@github.com> 于 2017年12月29日周五 14:38写道：

**Comment by ChiZelin** [comment]
@dantewang 
https://github.com/liferay/liferay-portal/blob/1111b038424447878583504a4ae83ddf09a164d3/definitions/web-app_2_4.xsd#L1101

**Comment by dantewang** [comment]
@ChiZelin Found this from http://www.oracle.com/webfolder/technetwork/jsc/xml/ns/j2ee/web-app_2_4.xsd:

> When there are multiple elements of welcome-file-list or locale-encoding-mapping-list, the container must concatinate the element contents.

So let's keep the for-loop, but put all the mappings in those `local-encoding-mapping-list` elements into a single map, just create the config and map object outside the loop.

For the field name, let's just use the pluto one.

**Comment by ChiZelin** [comment]
@dantewang OK! Thanks!

**Comment by ChiZelin** [comment]
Please see https://github.com/dantewang/liferay-portal/pull/537

**Pull request closed without merging**